### PR TITLE
orchestrator: capture initial sketch

### PR DESCRIPTION
**Pull request by benesch**
This commit teaches `materialized` to optionally manage its own compute
clusters. This functionality is provided by two new crates:

  * The `orchestrator` crate provides an abstraction for managing
    services, which are a set of one or more processes running the same
    image.

  * The `orchestrator-kubernetes` crate provides a concrete
    implementation of the `Orchestrator` trait that is backed by
    Kubernetes. A service in the `KubernetesOrchestrator` corresponds to
    a StatefulSet in Kubernetes.

In the future, we will likely provide additional orchestrator
implementations that manage processes on the local machine or images in
a Docker daemon to ease local development. This commit adds only the
Kubernetes-backed implementation since that is what we plan to use in
production.

Local development of the orchestrator is facilitated by the addition of
a family of scripts which build and deploy `materialized` to Minikube.
The scripts are:

  * `bin/cluster-dev`, which deploys `materialized` into Minikube,
    essentially simulating the environment controller locally
  * `bin/cluster-psql`, which launches a psql shell connected to the
    `materialized` instance running in Minikube
  * `bin/cluster-reset`, which blows away all Minikube state

This is prototype quality code. There are many missing retry loops, and
the `materialized` that gets deployed to Kubernetes crash loops several
times before it becomes healthy. And there are no tests. But none of
this code is exercised in the default configuration, and using it is
gated by `--experimental`.

### Motivation

  * This PR adds a known-desirable feature.

### Testing

- [x] This PR has adequate test coverage / QA involvement has been duly considered: no tests here, because it's experimental code. Tests will be added later.

### Release notes

This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):

  - N/A (experimental)

### CODE REVIEW

**Comment by benesch** [comment]
Well I'll be damned. This pretty much works. If you want to try it out:

```shell
# Install and start Minikube.
$ brew install minikube
$ minikube start

# Build `materialized` locally and push it into minikube.
$ bin/cluster-dev

# Open the Minikube dashboard and wait for all the services to go green.
$ minikube dashboard

# Alternatively, use k9s...
$ brew install k9s
$ k9s

# Start psql in the cluster. You may need to run this twice.
$ bin/cluster-psql

# Start a new managed cluster.
=> CREATE CLUSTER foo SIZE 'small';
```

There's a big defect right now, which is that DNS resolution isn't immediate, and so Materialize will crash when it goes to resolve the hostnames of the storage/compute instances right after it creates them. Hilariously this eventually sort of lurches its way towards success, as eventually DNS sorts itself out, and if you get lucky the processes crash loop in exactly the right way and get you a working system... until they crash again.

**Comment by antiguru** [comment]
I had to apply the following patch to make this work on Linux. When building the materialized image, we mount docker volumes into the image. For minikube, this mounts the volumes *within minikube* into docker. This fails because the directories aren't mounted into minikube and hence it fails to compile as the sources aren't available.

The patch changes `ci-builder` to not use minikube's docker daemon but directly the default. I think this is not a proper change as it just replaces a configured daemon with the default, not taking into account a different configuration that might apply to the host system.

```diff
diff --git a/bin/ci-builder b/bin/ci-builder
index 6d34f0c8b..949691417 100755
--- a/bin/ci-builder
+++ b/bin/ci-builder
@@ -204,6 +204,7 @@ case "$cmd" in
             )
         fi
         rm -f "$cid_file"
+        unset DOCKER_TLS_VERIFY DOCKER_HOST DOCKER_CERT_PATH MINIKUBE_ACTIVE_DOCKERD
         docker run "${args[@]}" "materialize/ci-builder:$tag" "$@"
         ;;
     root-shell)
```

File: src/dataflow-types/src/client/controller.rs [hunk 1]
```diff
@@ -40,12 +42,30 @@ mod storage;
...
+    /// Run a local storage instance.
+    Local,
+    /// Use a remote storage instance.
+    Remote {
+        /// The address that compute instances should connect to.
+        compute_addr: String,
+        /// The address that the controller should connect to.
+        controller_addr: String,
+        /// The number of workers in the remote process.
+        workers: usize,
```
Comment (antiguru) [inline src/dataflow-types/src/client/controller.rs:57]: This configuration param .. we could have the tcp_boundary communicate the number of workers from the server to the client and not specify it on the client.

File: src/dataflow-types/src/client/controller.rs [hunk 2]
```diff
@@ -71,6 +91,53 @@ where
...
+                    .namespace("compute")
+                    .ensure_service(
+                        &format!("instance-{instance}"),
+                        ServiceConfig {
+                            image: "materialize/dataflowd:unstable".into(),
+                            args: vec![
+                                "--runtime=compute".into(),
+                                format!("--storage-workers={storage_workers}"),
+                                format!("--storage-addr={storage_compute_addr}"),
+                                "0.0.0.0:2101".into(),
```
Comment (antiguru) [inline src/dataflow-types/src/client/controller.rs:120]: Not sure what the k8s conventions are, but hardcoded IPs/ports seem dangerous to me!

**Review by antiguru — APPROVED** [review]
LGTM

File: src/dataflow-types/src/client/controller.rs (hunk 1 above)
Comment (antiguru) [reply src/dataflow-types/src/client/controller.rs:57]: I'll file an issue for this. Don't bother here!

Comment (antiguru) [reply src/dataflow-types/src/client/controller.rs:57]: https://github.com/MaterializeInc/materialize/issues/11355

Comment (benesch) [reply src/dataflow-types/src/client/controller.rs]: Yes! Thanks @antiguru!

**Pull request merged**